### PR TITLE
[WIP] Kubernetes integration

### PR DESCRIPTION
**Pull request by Zogg**
[Work in Progress]
Kubernetes integration into Mantl. This is an integration of [kubernetes-ansible](https://github.com/CiscoCloud/kubernetes-ansible).

## DNS
We run etcd/skydns on every node. DNS entries are managed by kube2sky. DNS queries are routed through dnsmasq>skydns>etcd on local node to decrease dns network traffic.

## Networking
Kubernetes may employ either Flannel or Calico for containter networking.

### Calico
* Pitfall: Currently calico cannot assign IP to container/pod which has been launched from manifest file. Only pods launched through kalicoctl/kubeapi gets an IP assigned. Track the bug in [kubernetes repo](https://github.com/kubernetes/kubernetes/issues/14992) and [calico-kubernetes repo](https://github.com/projectcalico/calico-kubernetes/issues/55).

### Flanneld
* Blocker: flanneld cannot probperly start because of race condition between services. Flannel needs to start before Docker, Etcd must be started before Flannel, but currently it is started as Docker container. Hence the lockdown. Track the [issue](https://github.com/CiscoCloud/microservices-infrastructure/issues/795).

### CODE REVIEW

**Comment by stevendborrelli** [comment]
@zogg what is the impact of the calico issue? How often do we deploy via manifest vs. the API.

**Comment by siddharthist** [comment]
Related: #643

**Comment by ohal** [comment]
@jenkins please test it

**Comment by ohal** [comment]
@jenkins ok to test

**Comment by eirslett** [comment]
Is it possible to run Kubernetes on Mantl, on top of Mesos instead of using the bundled Kubernetes scheduler? (Using the Kubernetes mesos framework)
In which case, it could probably be run from Marathon instead of as a system process?

**Comment by stevendborrelli** [comment]
@eirslett The general feeling is that it is less complex to run kubernetes on its own than by using the k8s-mesos framework. Mantl support for k8s experimental at this point, and we are going to keep it separate from Mesos installs until we can integrate it better. 

The idea is to migrate K8s to use calico and consul, so that we have common networking and service discovery for different frameworks.

File: roles/calico/defaults/main.yml [hunk 1]
```diff
@@ -1,14 +1,14 @@
 ---
-calico_image: ciscocloud/calico-node
-calico_image_tag: v0.4.10-mi
+calico_image: calico/node
+calico_image_tag: v0.9.0
 
-bintray_baseurl: "https://bintray.com/artifact/download/ciscocloud/generic"
-calico_binary: "{{ bintray_baseurl }}/calico/calicoctl"
+bintray_baseurl: "https://bintray.com/artifact/download/zogg"
```
Comment (ryane) [inline roles/calico/defaults/main.yml]: could we use an official package for this or move it to ciscocloud?

File: roles/calico/tasks/main.yml [hunk 2]
```diff
@@ -39,14 +39,14 @@
...
 
-- name: download calicoctl bash completion script
-  sudo: yes
-  get_url:
-    url: "{{ bintray_baseurl }}/calico/calicoctl.sh"
-    dest: /etc/bash_completion.d/calicoctl.sh
-    mode: 0644
-  tags:
-    - calico
+#- name: download calicoctl bash completion script
```
Comment (ryane) [inline roles/calico/tasks/main.yml]: should we just remove this?

File: kubernetes_vars.yml [hunk 3]
```diff
@@ -0,0 +1,76 @@
+# ansible remote user account
+# ansible_ssh_user: centos
+
+# ansible groups for service configuration
+etcd_group_name: role=control
```
Comment (ryane) [inline kubernetes_vars.yml]: can we add this variable to the roles/etcd/defaults/main.yml so that etcd can be included without kubernetes if needed?

**Comment by ryane** [comment]
I was able to successfully run the kubernetes playbook. 

but the changes in the etcd and calico roles make them unusable on a non-kubernetes cluster.

**Comment by stevendborrelli** [comment]
@Zogg @avnik we need to get the etcd playbook aligned so the same one is used for kubernetes and calico.

File: roles/calico/defaults/main.yml (hunk 1 above)
Comment (Zogg) [inline roles/calico/defaults/main.yml]: Yes, I'm on to it.

File: kubernetes_vars.yml (hunk 3 above)
Comment (Zogg) [inline kubernetes_vars.yml]: Yes. Done.

File: roles/calico/defaults/main.yml (hunk 1 above)
Comment (Zogg) [inline roles/calico/defaults/main.yml]: Moved into asteris account.

**Comment by Zogg** [comment]
@stevendborrelli The roles are fixed and OK.

**Comment by ryane** [comment]
etcd playbook does seem to work well in a non-kubernetes mantl cluster. and survives restarts so this fixes #737. 

calico is still problematic on a non-kubernetes cluster. no containers can be launched with mesos/marathon when calico is enabled.

**Comment by ryane** [comment]
we're going to address the calico problems in a separate issue

**Pull request merged**